drivers/usb/udc/udc_mcux_ehci.c

```diff
@@ -0,0 +1,920 @@
+{
+	return udc_lock_internal(dev, K_FOREVER);
+}
+
+static int usb_mcux_unlock(const struct device *dev)
+{
+	return udc_unlock_internal(dev);
+}
+
+static int usb_mcux_driver_preinit(const struct device *dev)
+{
+	const struct udc_mcux_config *config = dev->config;
+	struct udc_data *data = dev->data;
+	struct udc_mcux_data *priv = data->priv;
+	int err;
+	/* MCUX USB controller drivers use one ID to tell controller
+	 * drivers which controller is used. This part of codes are
+	 * convert the IP base address to the ID value.
+	 */
+	uint8_t index;
+#ifdef USBHS_STACK_BASE_ADDRS
+	uint32_t usb_base_addrs[] = USBHS_STACK_BASE_ADDRS;
+#else
+	uint32_t usb_base_addrs[] = USBHS_BASE_ADDRS;
+#endif
+
+	/* get the right controller id */
+	priv->controller_id = 0xFFu; /* invalid value */
+	for (index = 0; index < ARRAY_SIZE(usb_base_addrs); index++) {
+		if (usb_base_addrs[index] == config->base) {
+			priv->controller_id = kUSB_ControllerEhci0 + index;
+			break;
+		}
+	}
+
+	if (priv->controller_id == 0xFFu) {
+		return -ENOMEM;
+	}
+
+	k_mutex_init(&data->mutex);
+
+	for (int i = 0; i < config->num_of_eps; i++) {
+		config->ep_cfg_out[i].caps.out = 1;
+		if (i == 0) {
+			config->ep_cfg_out[i].caps.control = 1;
+			config->ep_cfg_out[i].caps.mps = 64;
+		} else {
+			config->ep_cfg_out[i].caps.bulk = 1;
+			config->ep_cfg_out[i].caps.interrupt = 1;
+			config->ep_cfg_out[i].caps.iso = 1;
+			config->ep_cfg_out[i].caps.mps = 1024;
+		}
+
+		config->ep_cfg_out[i].addr = USB_EP_DIR_OUT | i;
+		err = udc_register_ep(dev, &config->ep_cfg_out[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	for (int i = 0; i < config->num_of_eps; i++) {
+		config->ep_cfg_in[i].caps.in = 1;
+		if (i == 0) {
+			config->ep_cfg_in[i].caps.control = 1;
+			config->ep_cfg_in[i].caps.mps = 64;
+		} else {
+			config->ep_cfg_in[i].caps.bulk = 1;
+			config->ep_cfg_in[i].caps.interrupt = 1;
+			config->ep_cfg_in[i].caps.iso = 1;
+			config->ep_cfg_in[i].caps.mps = 1024;
+		}
+
+		config->ep_cfg_in[i].addr = USB_EP_DIR_IN | i;
+		err = udc_register_ep(dev, &config->ep_cfg_in[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	data->caps.rwup = true;
+	data->caps.mps0 = USB_MCUX_MPS0;
+	data->caps.hs = true;
+	priv->dev = dev;
+
+	pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+
+	return 0;
+}
+
+static const struct udc_api usb_mcux_api = {
+	.device_speed = udc_mcux_device_speed,
+	.ep_enqueue = usb_mcux_ep_enqueue,
+	.ep_dequeue = usb_mcux_ep_dequeue,
+	.ep_set_halt = usb_mcux_ep_set_halt,
+	.ep_clear_halt = usb_mcux_ep_clear_halt,
+	.ep_try_config = NULL,
+	.ep_enable = usb_mcux_ep_enable,
+	.ep_disable = usb_mcux_ep_disable,
+	.host_wakeup = usb_mcux_host_wakeup,
+	.set_address = usb_mcux_set_address,
+	.enable = usb_mcux_enable,
+	.disable = usb_mcux_disable,
+	.init = usb_mcux_init,
+	.shutdown = usb_mcux_shutdown,
+	.lock = usb_mcux_lock,
+	.unlock = usb_mcux_unlock,
```

Comment:
Should the prefix be usb_mcux? I suggest to prefix it consistently with udc_mcux.

Reply:
Updated.